### PR TITLE
ivreg methods

### PR DESCRIPTION
**Pull request by nathanielolin**
Implements margins for `AER::ivreg` that replicate Stata's `ivregress 2sls` method.

### CODE REVIEW

**Comment by nathanielolin** [comment]
TODOs include testing weights, missingness, and combos thereof, but the core is here.

File: R/ivreg_marg.R [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+#' @importFrom stats get_all_vars
+get_data.ivreg <- function(model, data, weights){
+
+  # Store original number of rows
+  nrow_orig <- nrow(data)
+
+  # Grab only necessary variables
+  data <- data[, all.vars(model$formula)]
+
+  # Add weights
```
Comment (anniejw6) [inline R/ivreg_marg.R]: Everything from here on below seems similar to `glm`. Can we create a helper function and change both of those?

File: R/ivreg_marg.R [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    warning(sprintf('Dropping %s rows due to missing data',
+                    nrow_orig - nrow(data)))
+
+  list(data = data,
+       weights = weights)
+
+}
+
+# Method taken from unexported AER::vcov.ivreg method
+# (https://github.com/cran/AER/blob/master/R/ivreg.R)
```
Comment (anniejw6) [inline R/ivreg_marg.R]: Can you add link to specific line numbers via permalink?

File: R/ivreg_marg.R [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+#' # Fit the model
+#' mod <- AER::ivreg(
+#'   y ~ as.factor(actual) + age + distance |
+#'     as.factor(assign) + age + distance,
+#'   data = margex)
+#'
+#' # Get margins for different levels of "actual" treatment
+#' modmarg::marg(mod, data = margex, var_interest = 'actual')
+#'
+#' # TODO: weights, clustering
```
Comment (anniejw6) [inline R/ivreg_marg.R]: This can be a separate MR

File: R/ivreg_marg.R [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+#'
+#' Obtains predicted margins and standard errors
+#' of those predictions from a fitted \code{ivreg} model object.
+#'
+#' @inheritParams marg
+#' @param ... additional parameters passed to \code{?marg}.
+#'
+#' @export
+#' @examples
+#' data(margex)
```
Comment (anniejw6) [inline R/ivreg_marg.R]: Why doesn't this example just load in some data from the `AER` package instead of going to such lengths to build our own?

File: R/ivreg_marg.R [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+  model$sigma^2 * model$cov.unscaled * model$df.residual / model$nobs
+}
+
+get_dof.ivreg <- function(model, ...){
+  Inf
+}
+
+#' @importFrom stats delete.response model.frame na.pass
+get_covar.ivreg <- function(model, data){
+
```
Comment (anniejw6) [inline R/ivreg_marg.R]: Can you add some comments to be clear about what's going on here?

**Comment by nathanielolin** [comment]
If the builds pass, this is good to review again.

File: R/ivreg_marg.R [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+#' @importFrom stats get_all_vars
+get_data.ivreg <- function(model, data, weights){
+
+  # Store original number of rows
+  nrow_orig <- nrow(data)
+
+  # Grab only necessary variables
+  data <- data[, all.vars(model$formula)]
```
Comment (anniejw6) [inline R/ivreg_marg.R]: Should this have a `, drop`? I know it's kinda silly because there has to be more than one covar, but ... maybe?

**Comment by nathanielolin** [comment]
Test added - back to @anniejw6

File: tests/testthat/test_ivreg.R [hunk 7]
```diff
@@ -90,6 +94,11 @@ test_that('2SLS margins are correct', {
   expect_equal(z$Label, as.factor(paste('actual =', c(0, 1))))
   expect_equal(z$Margin, c(61.84719, 91.4581), tolerance = 0.000001)
   expect_equal(z$Standard.Error, c(.6426773, 1.417176), tolerance = 0.0000001)
+  # TODO: other parameters?
```
Comment (anniejw6) [inline tests/testthat/test_ivreg.R]: Given our past challenges with p-values, et.c, can we test those too?

Comment (anniejw6) [reply tests/testthat/test_ivreg.R]: I know it's in there later, but those are all places where it's zero...

Comment (anniejw6) [reply tests/testthat/test_ivreg.R]: I guess it's implictly handled by the conf intervals...?

**Comment by nathanielolin** [comment]
@anniejw6 added test for p values for models fit on random noise.

File: tests/testthat/test_ivreg.R (hunk 7 above)
Comment (anniejw6) [reply tests/testthat/test_ivreg.R]: Resolved
